src/Microsoft.Diagnostics.EventFlow.Outputs.EventHub/EventHubOutput.cs

```diff
@@ -83,52 +84,51 @@ public EventHubOutput(IConfiguration configuration, IHealthReporter healthReport
 
             try
             {
-                // Since event hub limits each message/batch to be a certain size, we need to
-                // keep checking the size for exceeds and split into a new batch as needed
+                var groupedEventData = events.GroupBy(e => string.IsNullOrEmpty(outputConfiguration.PartitionKeyProperty) == false && e.TryGetPropertyValue(outputConfiguration.PartitionKeyProperty, out var partionKeyData) ? partionKeyData.ToString() : string.Empty, e => e);
 
-                List<List<MessagingEventData>> batches = new List<List<MessagingEventData>>();
-                int batchByteSize = 0;
+                List<Task> tasks = new List<Task>();
 
-                foreach (EventData eventData in events)
+                foreach (var partitionedEventData in groupedEventData)
                 {
-                    int messageSize;
-                    MessagingEventData messagingEventData = eventData.ToMessagingEventData(out messageSize);
+                    SendBatch(partitionedEventData.Select(e => new Tuple<MessagingEventData, int>(e.ToMessagingEventData(out var messageSize), messageSize)).ToList());
```

Comment:
If you use tuples, consider C# 7 build-in support https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7 that lets you name tuple members instead of using Item1, Item2 and so on.

Reply:
@karolz-ms I implemented named tuple members, thanks for this suggestion